src/SmartSignalsRuntimeShared/AzureStorage/BlobStateRepository.cs

```
@@ -0,0 +1,151 @@
+            }
+
+            key = key.ToLowerInvariant();
+
+            string serializedState = JsonConvert.SerializeObject(state);
+
+            BlobState blobState = new BlobState
+            {
+                Key = key,
+                SignalId = this.signalId,
+                SerializedState = serializedState
+            };
+
+            string serializedBlobState = JsonConvert.SerializeObject(blobState);
+
+            return this.cloudBlobContainerWrapper.UploadBlobAsync(this.GenerateBlobName(key), serializedBlobState, cancellationToken);
+        }
+
+        /// <summary>
+        /// Gets signal state by key
+        /// </summary>
+        /// <typeparam name="T">Type type of the state</typeparam>
+        /// <param name="key">The key (case insensitive)</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>A state associated with the signal and the key wrapped in a <see cref="Task{T}"/></returns>
+        public async Task<T> GetStateAsync<T>(string key, CancellationToken cancellationToken)
+        {
+            Diagnostics.EnsureArgumentNotNull(() => key);
+
+            key = key.ToLowerInvariant();
+
+            string serializedBlobState = null;
+            try
+            {
+                serializedBlobState = await this.cloudBlobContainerWrapper.DownloadBlobContentAsync(this.GenerateBlobName(key), cancellationToken);
+            }
+            catch (StorageException ex) when ((HttpStatusCode)ex.RequestInformation.HttpStatusCode == HttpStatusCode.NotFound)
+            {
+                this.tracer.TraceWarning("State not found in the repository, returning empty state");
+                return default(T);
+            }
+
+            BlobState blobState = JsonConvert.DeserializeObject<BlobState>(serializedBlobState);
+
+            if (string.Equals(blobState.SignalId, this.signalId) == false 
+                || string.Equals(blobState.Key, key) == false
```

This means something really wrong has happened - let's throw an exception in these 2 cases case, so we can fix this bug if it happens